### PR TITLE
feat!: Python 3.12 Upgrade

### PR DESCRIPTION
**Pull request by UsamaSadiq**
## Description
- Updated GitHub CI and tox to run checks with `Python 3.12` env only
- Updated requirements with `Python 3.12` env
- Updated Dockerfile to build with `Python 3.12`.
- Dropped support for `Python<3.12`

## Testing
- [x] Tested the docker build locally using `make_build` with updated Dockerfile to be successful.
- [x] CI checks with `Python 3.12` are green.

## TODO
- Python 3.8 checks marked as required need to be removed and Python 3.12 checks need to be marked as required before merging this PR.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -47,6 +51,9 @@ ENV DISCOVERY_CODE_DIR "${DISCOVERY_CODE_DIR}"
 ENV DISCOVERY_APP_DIR "${DISCOVERY_APP_DIR}"
 ENV PYTHON_VERSION "${PYTHON_VERSION}"
 
+# Setup zoneinfo for Python 3.12
+RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
```
Comment (DawoudSheraz) [inline Dockerfile:55]: Any reference on why this change is needed in 3.12?

**Review by DawoudSheraz — APPROVED** [review]
I have not tested it locally, looks alright. Please squash the commits.

File: Dockerfile (hunk 1 above)
Comment (UsamaSadiq) [reply Dockerfile:55]: This changes was needed in the enterprise IDAs when upgrading to Python 3.12 quite possibly because of new zoneinfo changes in Python 3.12 so I added it here as well to keep the changes similar across all services.

**Comment by UsamaSadiq** [comment]
> I have not tested it locally, looks alright. Please squash the commits.

Commits are squashed but we'll need to remove the required checks from repo settings as mentioned in the TODO section of PR description. I don't have access for that so someone from the team or SRE will have to do this change.

**Comment by DawoudSheraz** [comment]
> > I have not tested it locally, looks alright. Please squash the commits.
> 
> Commits are squashed but we'll need to remove the required checks from repo settings as mentioned in the TODO section of PR description. I don't have access for that so someone from the team or SRE will have to do this change.

Only Axim can do this as they have taken the admin-ship back for Open edX repositories from maintainers as well.

**Review by AfaqShuaib09 — APPROVED** [review]

**Comment by UsamaSadiq** [comment]
FYI @DawoudSheraz I've created axim-engineering issue to remove the required py38 checks from the repo settings. Would you be able to merge the PR after this or does only axim has merge access as well?

**Comment by DawoudSheraz** [comment]
> FYI @DawoudSheraz I've created axim-engineering issue to remove the required py38 checks from the repo settings. Would you be able to merge the PR after this or does only axim has merge access as well?

I can merge the PR once done.

**Comment by feanil** [comment]
Hey all, the Axim ticket was done and this is blocking some other work so I'm gonna merge this now.

**Comment by UsamaSadiq** [comment]
@feanil Thanks for merging this PR. Should I raise another ticket to mark `Python 3.12` checks as required?

**Comment by feanil** [comment]
I just added them.  FYI in the future, if you have a PR that has run the new checks you can add them even if master doesn't run those checks yet so you can actually make a ticket to remove the old ones and add the new ones if the PR with the new ones has run recently.
